contracts/governance/proposal/ProposalMasterCharlieV1.sol

```diff
@@ -0,0 +1,431 @@
+
+    /// @notice The governance master contract to retrieve metadata regarding
+    /// other governance aspects such as voting power.
+    IGovernor public immutable governance;
+
+    /// @dev The set of tokens used for governance purposes.
+    EnumerableSet.AddressSet private tokens;
+
+    mapping (address => EnumerableSet.AddressSet) private tokensByGovernance;
+
+    /// @dev The registry of addresses to exclude for the purpose of counting
+    /// actual votes. Examples can be burn addresses and treasuries.
+    mapping (address => EnumerableSet.AddressSet) private excluded;
+
+    /// @notice The percentage of circulating tokens to use as threshold for
+    /// proposal creation eligibility.
+    uint256 public proposalPercentage;
+
+    /// @notice The percentage of circulating tokens to use as threshold for
+    /// required quorum to pass a proposal.
+    uint256 public quorumPercentage;
+
+    /// @notice The period (expressed in blocks) before voting goes live for a
+    /// given proposal.
+    uint256 public proposalDelay;
+
+    /// @notice The period (expressed in blocks) in which voters are able to
+    /// cast a vote for a given proposal.
+    uint256 public proposalVotingPeriod;
+
+    /// @notice Emitted when the proposal threshold percentage is updated.
+    event ProposalPercentageUpdated(
+        uint256 oldPercentage,
+        uint256 newPercentage
+    );
+
+    /// @notice Emitted when the quorum threshold percentage is updated.
+    event QuorumPercentageUpdated(
+        uint256 oldPercentage,
+        uint256 newPercentage
+    );
+
+    /// @notice Emitted when a new governance token is added.
+    event TokenAdded(address token);
+
+    /// @notice Emitted when a governance token is removed.
+    event TokenRemoved(address token);
+
+    /// @notice Emitted when an account's balance is excluded from being counted
+    /// towards a token's total supply.
+    event ExclusionAccountAdded(address token, address account);
+
+    /// @notice Emitted when an account's balance is no longer excluded from
+    /// being counted towards a token's total supply.
+    event ExclusionAccountRemoved(address token, address account);
+
+    /**
+     * @param tokens_ The initial list of governance tokens to support.
+     * @param timelock The timelock address associated with governance.
+     * @param governance_ The master governance contract address.
+     * @param initialProposalPercentage The percentage to use as threshold for
+     * checking proposal creation eligibility.
+     * @param initialQuorumPercentage The percentage to use for calculating the
+     * required quorum for passing a proposal.
+     */
+    constructor(
+        address[] memory tokens_,
+        address timelock,
+        address governance_,
+        uint256 initialProposalPercentage,
+        uint256 initialQuorumPercentage
+    ) {
+        for (uint256 i = 0; i < tokens_.length; i++) {
+            addToken(tokens_[i]);
```

Comment:
don't you want to check for address(0) before adding?

Reply:
Fair point, wouldn't hurt. Will add.